utxo/src/undo.rs

```diff
@@ -100,20 +118,52 @@ impl BlockUndo {
     pub fn insert_tx_undo(
         &mut self,
         tx_id: Id<Transaction>,
-        tx_undo: TxUndo,
+        tx_undo: TxUndoWithSources,
     ) -> Result<(), BlockUndoError> {
         match self.tx_undos.entry(tx_id) {
-            std::collections::btree_map::Entry::Vacant(e) => e.insert(tx_undo),
-            std::collections::btree_map::Entry::Occupied(_) => {
-                return Err(BlockUndoError::UndoAlreadyExists(tx_id))
-            }
+            Entry::Vacant(e) => e.insert(TxUndo::new(tx_undo.utxos)),
+            Entry::Occupied(_) => return Err(BlockUndoError::UndoAlreadyExists(tx_id)),
         };
+
+        tx_undo
+            .sources
+            .into_iter()
+            .filter_map(|source_id| match source_id {
+                OutPointSourceId::Transaction(id) => Some(id),
+                OutPointSourceId::BlockReward(_) => None,
+            })
+            .for_each(|source_tx_id| {
+                self.child_parent_dependencies.insert((tx_id, source_tx_id));
+                self.parent_child_dependencies.insert((source_tx_id, tx_id));
+            });
+
         Ok(())
     }
 
     pub fn take_tx_undo(&mut self, tx_id: &Id<Transaction>) -> Option<TxUndo> {
-        // TODO(PR): this should analyze dependencies before taking an undo
-        self.tx_undos.remove(tx_id)
+        let range_start = (*tx_id, Id::<Transaction>::from(H256::zero()));
+        let range_end = (*tx_id, Id::<Transaction>::from(H256::repeat_byte(0xFF)));
+
+        // Check if the tx is a dependency for other txs.
+        let dependencies_count =
+            self.parent_child_dependencies.range(range_start..=range_end).count();
+        if dependencies_count == 0 {
+            // If not this tx can be taken and returned.
+            // But first, remove itself as a dependency of others.
+            let to_remove = self
+                .child_parent_dependencies
+                .range(range_start..=range_end)
+                .copied()
+                .collect::<Vec<_>>();
```

Comment:
I don't think you need to collect here... you could just use the iterator in the next step.

Reply:
Actually, I do, because otherwise, I would hold references to `child_parent_dependencies` and won't be able to modify it

Reply:
I see.

Reply:
Perhaps the `drain_filter` method can be used here, but it is still unstable.